### PR TITLE
feat: improve error handling

### PR DESCRIPTION
**Pull request by greged93**
Improve the current error handling, heavily inspired by the following reth code:
- [ETH RPC error codes](https://github.com/paradigmxyz/reth/blob/9e72cbf6b44fd195738c6d85f461304578c8156c/crates/rpc/rpc-types/src/eth/error.rs#L5)
- [Jsonrpsee ObjectError helper](https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/result.rs#L159)
- [From implementation](https://github.com/paradigmxyz/reth/blob/9e72cbf6b44fd195738c6d85f461304578c8156c/crates/rpc/rpc/src/eth/error.rs#L81)

Time spent on this PR: 0.5 day

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?
Error handling is a little messy and naming unclear.

Resolves: #NA

# What is the new behavior?
Error handling is regrouped into one file, with clear error codes.

- Renaming of `KakarotClientError` into `EthApiError`
- Regrouping of code into one file `errors.rs`
- Definition of clear error codes

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/core/src/client/helpers.rs [hunk 1]
```diff
@@ -90,10 +85,8 @@ pub fn ethers_block_id_to_starknet_block_id(block: EthBlockId) -> Result<Starkne
 
 /// # Errors
 ///
-/// TODO: Will return `KakarotClientError`..
-pub const fn ethers_block_number_to_starknet_block_id(
-    block: BlockNumberOrTag,
-) -> Result<StarknetBlockId, KakarotClientError> {
+/// TODO: Will return `EthApiError`..
+pub const fn ethers_block_number_to_starknet_block_id(block: BlockNumberOrTag) -> Result<StarknetBlockId, EthApiError> {
```
Comment (greged93) [inline crates/core/src/client/helpers.rs:89]: Can I remove this comment?

File: crates/core/src/client/errors.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    /// Failed to send transaction, See also <https://github.com/MetaMask/eth-rpc-errors/blob/main/src/error-constants.ts>
+    TransactionRejected = -32003,
+    /// Custom geth error code, <https://github.com/vapory-legacy/wiki/blob/master/JSON-RPC-Error-Codes-Improvement-Proposal.md>
+    ExecutionError = 3,
+    /// <https://eips.ethereum.org/EIPS/eip-1898>
+    InvalidInput = -32000,
+    /// Thrown when a block wasn't found <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1898.md>
+    /// > If the block is not found, the callee SHOULD raise a JSON-RPC error (the recommended
+    /// > error code is -32001: Resource not found).
+    ResourceNotFound = -32001,
```
[the remark this answers is not in the document]
Comment (greged93) [reply crates/core/src/client/errors.rs:23]: [Random order in the definition yes](https://github.com/paradigmxyz/reth/blob/9e72cbf6b44fd195738c6d85f461304578c8156c/crates/rpc/rpc-types/src/eth/error.rs#L5). I can for sure sort it, prefer by name or code? Personally code but can do either.

File: crates/core/src/client/helpers.rs [hunk 3]
```diff
@@ -329,10 +314,10 @@ pub fn create_default_transaction_receipt() -> TransactionReceipt {
 /// # Errors
 ///
 /// TODO: add error case message
```
Comment (ClementWalter) [inline crates/core/src/client/helpers.rs:316]: do we need to update this?

**Review by ClementWalter — APPROVED** [review]

File: crates/core/src/client/errors.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ClementWalter) [reply crates/core/src/client/errors.rs:23]: as you want

File: crates/core/src/client/client_api.rs [hunk 4]
```diff
@@ -1,126 +1,97 @@
...
-    ) -> Result<Bytes, KakarotClientError>;
+    ) -> Result<Bytes, EthApiError>;
 
     async fn call_view(
         &self,
         ethereum_address: Address,
         calldata: Bytes,
         starknet_block_id: StarknetBlockId,
-    ) -> Result<Bytes, KakarotClientError>;
+    ) -> Result<Bytes, EthApiError>;
```
Comment (Eikix) [inline crates/core/src/client/client_api.rs:44]: I feel like `EthApiError` may be a confusing name. Here for instance, we call Starknet with a "call" request, it feels weird that we may return an `EthApiError`.

Could we name it `KakarotError`? Or something else, able to tell people this error comes from the core business logic of Kakarot.

Comment (Eikix) [reply crates/core/src/client/client_api.rs:44]: Alternatively, we could split our core business logic into:
StarknetWrapper trait and
KakarotEthApi trait, and then have KakarotClient be a composition of both traits,

Then, you can have EthApiError for KakarotEthApi trait, that only implements functions that have ~almost the same signature as Ethereum JSON RPC client calls.

Then, we can have StarknetWrapperError for StarknetWrapper trait, that only implements function that wrap starknet calls but format things to adapt to Kakarot business logic.

File: crates/core/src/client/helpers.rs (hunk 3 above)
Comment (greged93) [reply crates/core/src/client/helpers.rs:314]: Made [an issue](https://github.com/kkrt-labs/kakarot-rpc/issues/197) for this (Not so much the TODO but think we need to implement `TryFrom`).

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**